### PR TITLE
Replace StorageV1 client with GCS client (take 4)

### PR DESCRIPTION
**Pull request by BjornPrime**
Addresses #25676.

Now with improved error handling on batch requests (and test coverage of same).

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/io/gcp/gcsio.py [hunk 1]
```diff
@@ -271,186 +190,110 @@ def delete(self, path):
...
+          src_bucket_name, src_blob_name = parse_gcs_path(pair[0])
+          dest_bucket_name, dest_blob_name = parse_gcs_path(pair[1])
+          src_bucket = self.client.get_bucket(src_bucket_name)
+          src_blob = src_bucket.get_blob(src_blob_name)
+          dest_bucket = self.client.get_bucket(dest_bucket_name)
+
+          src_bucket.copy_blob(src_blob, dest_bucket, dest_blob_name)
 
-    return [(pair[0], pair[1], pair_to_status[pair]) for pair in src_dest_pairs]
+      for pair, resp in list(zip(current_pairs, current_batch._responses)):
```
[the remark this answers is not in the document]
Comment (BjornPrime) [reply sdks/python/apache_beam/io/gcp/gcsio.py:289]: I'm pretty sure I'm testing for this already, but I've added an assertion just to be sure

[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/python/apache_beam/io/gcp/gcsio.py:289]: I made a test like this

https://github.com/Abacn/beam/blob/4a84d1618ed23614ec663f2cda1758d280ed5da6/sdks/python/apache_beam/io/gcp/gcsio_test.py#L481-500 

also added some print to gcsio.py

https://github.com/Abacn/beam/blob/4a84d1618ed23614ec663f2cda1758d280ed5da6/sdks/python/apache_beam/io/gcp/gcsio.py#L292

Output:
```
5 20
200 200 200 200
200 200 200 200
200 200 200 200
200 404 200 404
200 200 200 200
gs://clouddfe-yihu-test/temp1/gcs_0.txt gs://clouddfe-yihu-test/temp2/gcs_0.txt 200
gs://clouddfe-yihu-test/temp1/gcs_1.txt gs://clouddfe-yihu-test/temp2/gcs_1.txt 200
gs://clouddfe-yihu-test/temp1/gcs_2.txt gs://clouddfe-yihu-test/temp2/gcs_2.txt 200
gs://clouddfe-yihu-test/temp1/gcs_3.txt gs://clouddfe-yihu-test/temp2/gcs_3.txt 200
gs://clouddfe-yihu-test/temp1/gcs_4.txt gs://clouddfe-yihu-test/temp2/gcs_4.txt 200
```
As can be seen, the length current_pairs and current_batch._responses is different (4 fold), and the return value missed the 404 code

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

File: sdks/python/apache_beam/io/gcp/gcsio.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BjornPrime) [reply sdks/python/apache_beam/io/gcp/gcsio.py:289]: I see what you're saying, and running the tests with the assertion included has confirmed it. I'll dig into the client more to see what we can do about it.

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python_Runners PreCommit 3.10

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

**Comment by BjornPrime** [comment]
Run Python 3.11 PostCommit

File: sdks/python/apache_beam/options/pipeline_options_validator_test.py [hunk 2]
```diff
@@ -106,6 +106,7 @@ def test_missing_required_options(self):
                                         ]), ('gs://foo', 'gs://foo/bar', []),
       ('gs://foo/', 'gs://foo/bar', []), ('gs://foo/bar', 'gs://foo/bar', [])
   ])
+  @unittest.skip('Not compatible with new GCS client. See GH issue #26335.')
```
Comment (Abacn) [inline sdks/python/apache_beam/options/pipeline_options_validator_test.py:109]: While it was understandable some bigquery unit test (related to file_load) was not compatible with new gcs client (they need adjust mock). How could pipeline option validator test be affected by the change?

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 3]
```diff
@@ -492,12 +491,19 @@ def __init__(self, options, root_staging_location=None):
     self._root_staging_location = (
         root_staging_location or self.google_cloud_options.staging_location)
 
-    self.environment_version = _FNAPI_ENVIRONMENT_MAJOR_VERSION
+    from apache_beam.runners.dataflow.dataflow_runner import _is_runner_v2_disabled
+    from google.cloud import storage
+    if _is_runner_v2_disabled(options):
```
Comment (Abacn) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:496]: Seems irrelevant change. Rebase error? Specifically it undo [this change](https://github.com/apache/beam/commit/ca674aad086e4d1a40e8c01598073030a22484fa#diff-e616be01292a173f978dab1922575bbe06c2554b3bc6073de4aa1117b81f5a13) on master

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 4]
```diff
@@ -654,38 +658,45 @@ def stage_file(
...
+            'that credentials are valid, that the specified path '
+            'exists, and that you have write access to it.') %
+                      (gcs_or_local_path, e))
       raise
-    _LOGGER.info(
-        'Completed GCS upload to %s in %s seconds.',
-        gcs_location,
-        int(time.time() - start_time))
-    return response
+    finally:
```
Comment (Abacn) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:695]: "finally" will execute when there are errors, at this point the message "Completed GCS upload" is not accurate

File: sdks/python/setup.py [hunk 5]
```diff
@@ -308,30 +308,31 @@ def get_portability_package_data():
...
-              'google-cloud-bigtable>=2.19.0,<3',
-              'google-cloud-spanner>=3.0.0,<4',
-              # GCP Packages required by ML functionality
-              'google-cloud-dlp>=3.0.0,<4',
-              'google-cloud-language>=2.0,<3',
-              'google-cloud-videointelligence>=2.0,<3',
-              'google-cloud-vision>=2,<4',
-              'google-cloud-recommendations-ai>=0.1.0,<0.11.0',
-              'google-cloud-aiplatform>=1.26.0, < 2.0'
+            'cachetools>=3.1.0,<6',
```
Comment (Abacn) [inline sdks/python/setup.py:311]: please fix diff (unnecessary indent change)

File: sdks/python/apache_beam/io/gcp/gcsio.py [hunk 6]
```diff
@@ -271,186 +190,119 @@ def delete(self, path):
...
+        for j in range(4):
+          resp = current_batch._responses[4 * i + j]
+          if resp.status_code >= 400:
+            error_code = resp.status_code
+            break
+        final_results.append((pair[0], pair[1], error_code))
+
+      s += MAX_BATCH_OPERATION_SIZE
+
+    return final_results
```
Comment (Abacn) [inline sdks/python/apache_beam/io/gcp/gcsio.py:302]: With the current code logic the "Returns:" spec is no longer true. Now it only returns a list of failed items. For compatibility reason I suggest we still return succeeded also item and keep "List of tuples of (src, dest, exception) in the same order as the src_dest_pairs argument" the spec.

Same as delete_batch

**Review by Abacn — COMMENTED** [review]
Thanks for persistent on this. I went through carefully and commented on potential issues I found.

File: sdks/python/apache_beam/io/gcp/bigquery_tools_test.py [hunk 7]
```diff
@@ -224,23 +223,6 @@ def test_delete_dataset_retries_for_timeouts(self, patched_time_sleep):
     wrapper._delete_dataset('', '')
     self.assertTrue(client.datasets.Delete.called)
 
-  @unittest.skipIf(
```
[the remark this answers is not in the document]
Comment (BjornPrime) [reply sdks/python/apache_beam/io/gcp/bigquery_tools_test.py:227]: I discussed this with Pablo and we agreed that the test was no longer necessary. While its failure is curious, IIRC, we determined that the test was not ultimately testing something relevant to production, so we didn't actually need to worry about it.

File: sdks/python/apache_beam/options/pipeline_options_validator_test.py (hunk 2 above)
Comment (BjornPrime) [reply sdks/python/apache_beam/options/pipeline_options_validator_test.py:109]: Tough to say for sure. My theory is that the dummy project name is maybe more of an issue for GCS than for apitools. The reason I created the GH issue is so that can be looked into more closely in the future.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py (hunk 3 above)
Comment (BjornPrime) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:496]: It does. Not sure how that happened. If we're no longer supporting the legacy runner (as I seem to recall we are not), I'll make sure to change it.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py (hunk 4 above)
Comment (BjornPrime) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:695]: Good catch, I'll fix that

File: sdks/python/setup.py (hunk 5 above)
Comment (BjornPrime) [reply sdks/python/setup.py:311]: Don't know how that happened but I'll resolve it

File: sdks/python/apache_beam/io/gcp/gcsio.py (hunk 6 above)
Comment (Abacn) [reply sdks/python/apache_beam/io/gcp/gcsio.py:302]: ah sorry you are correct

File: sdks/python/apache_beam/io/gcp/bigquery_tools_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/python/apache_beam/io/gcp/bigquery_tools_test.py:227]: This tests if user agent info is added to the request sent to bigquery. In various gcp component we use this info to track usage. This effort is active e.g. gcs counterpart #26770 #27211

That's possibly why it is saying not relevent to production, but does not mean it can be removed

We can make it "skipped" but please do not remove test

[the remark this answers is not in the document]
Comment (BjornPrime) [reply sdks/python/apache_beam/io/gcp/bigquery_tools_test.py:227]: Okay, I'll skip it and open an issue for it, since you say it's important.

**Comment by Abacn** [comment]
Run Python 3.11 PostCommit

**Review by Abacn — APPROVED** [review]

**Comment by Abacn** [comment]
PreCommit Portable Python workflow fails during clean up unrelated. Opened https://github.com/actions/setup-java/issues/543
